polymetis/polymetis/python/polysim/envs/mujoco_manipulator.py

```diff
@@ -0,0 +1,198 @@
+            mujoco.glfw.glfw.swap_interval(1)
+
+            mujoco.mjv_defaultCamera(self.gui_camera)
+            mujoco.mjv_defaultOption(self.gui_opt)
+
+            self.gui_scene = mujoco.MjvScene(self.robot_model, maxgeom=10000)
+            self.gui_context = mujoco.MjrContext(
+                self.robot_model, mujoco.mjtFontScale.mjFONTSCALE_150.value
+            )
+
+    def reset(self, joint_pos: List[float] = None, joint_vel: List[float] = None):
+        """Reset the environment."""
+        mujoco.mj_resetData(self.robot_model, self.robot_data)
+        if joint_pos is None:
+            self.robot_data.qpos = self.rest_pose
+        if joint_vel is not None:
+            self.robot_data.qvel = joint_vel
+
+    def get_num_dofs(self) -> int:
+        """Get the number of degrees of freedom for controlling the simulation.
+
+        Returns:
+            int: Number of control input dimensions
+        """
+        return self.n_dofs
+
+    def get_current_joint_pos_vel(self) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Returns:
+            np.ndarray: Joint positions
+            np.ndarray: Joint velocities
+        """
+        return (
+            self.robot_data.qpos,
+            self.robot_data.qvel,
+        )
+
+    def get_current_joint_torques(
+        self,
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+        """
+        Returns:
+            np.ndarray: Torques received from apply_joint_torques
+            np.ndarray: Torques sent to robot (e.g. after clipping)
+            np.ndarray: Torques generated by the actuators (e.g. after grav comp)
+            np.ndarray: Torques exerted onto the robot
+        """
+        return (
+            self.prev_torques_commanded,
+            self.prev_torques_applied,
+            self.prev_torques_measured,
+            self.prev_torques_external,  # zeros
+        )
+
+    def apply_joint_torques(self, torques: np.ndarray):
+        """
+        input:
+            np.ndarray: Desired torques
+        Returns:
+            np.ndarray: final applied torque
+        """
+        self.prev_torques_commanded = torques
+
+        applied_torques = np.clip(torques, -self.torque_limits, self.torque_limits)
+
+        if self.use_grav_comp:
+            joint_cur_pos = self.get_current_joint_pos_vel()[0]
+            grav_comp_torques = self.compute_inverse_dynamics(
+                joint_pos=joint_cur_pos,
+                joint_vel=[0] * self.n_dofs,
+                joint_acc=[0] * self.n_dofs,
+            )  # zero vel + acc to find gravity
+            applied_torques += grav_comp_torques
+        self.prev_torques_measured = applied_torques.copy()
+
+        self.robot_data.ctrl = applied_torques
+        mujoco.mj_step(self.robot_model, self.robot_data)
+
+        if self.gui:
+            self.render()
+
+        return applied_torques
+
+    def compute_inverse_dynamics(
```

Comment:
Quite important to unit-test this method (though you should test the others as well) by ensuring result is comparable (within epsilon) against e.g. Pinocchio or PyBullet output.